tokio-postgres/src/passfile/mod.rs

```diff
@@ -0,0 +1,192 @@
+                }
+            }
+            if allow_eol {
+                Ok(value)
+            } else {
+                Err(())
+            }
+        };
+
+        Ok(PassfileEntry {
+            hostname: parse_one_field(false)?,
+            port: parse_one_field(false)?,
+            dbname: parse_one_field(false)?,
+            user: parse_one_field(false)?,
+            password: parse_one_field(true)?,
+        })
+    }
+}
+
+fn field_matches(search_value: &[u8], file_value: &[u8]) -> bool {
+    if file_value == [b'*'] {
+        return true;
+    }
+    file_value == search_value
+}
+
+/// Removes trailing CR and/or LF from a string.
+///
+/// Intended to match libpq's pg_strip_crlf().
+fn strip_crlf(bb: &[u8]) -> &[u8] {
+    for (idx, b) in bb.iter().copied().enumerate().rev() {
+        if b != b'\n' && b != b'\r' {
+            return &bb[..=idx];
+        }
+    }
+    &bb[..0]
+}
+
+/// Searches passfile text for a match.
+///
+/// Intended to match libpq's behavior closely.
+/// If there is an IO error, returns None.
+async fn password_for_key<T>(key: &PassfileKey<'_>, buf: T) -> Option<Vec<u8>>
+where
+    T: tokio::io::AsyncBufReadExt + Unpin,
+{
+    let mut it = buf.split(b'\n');
+    // If we get an io error, just stop reading
+    while let Ok(Some(line)) = it.next_segment().await {
+        if line.starts_with(&[b'#']) {
+            continue;
+        }
+        let line = strip_crlf(&line);
+        if line.is_empty() {
+            continue;
+        }
+        if let Ok(entry) = PassfileEntry::new(line) {
+            if field_matches(key.hostname, &entry.hostname)
+                && field_matches(&key.port, &entry.port)
+                && field_matches(key.dbname, &entry.dbname)
+                && field_matches(key.user, &entry.user)
+            {
+                if entry.password.is_empty() {
+                    // To be consistent with libpq, in this case we need to
+                    // stop searching the password file, but not attempt to
+                    // use the empty password string.
+                    return None;
+                } else {
+                    return Some(entry.password);
+                }
+            }
+        }
+    }
+    None
+}
+
+/// Searches a passfile, returning the password from the first matching line.
+///
+/// Returns None if:
+///  - there is no match
+///  - the passfile doesn't exist
+///  - there is an error reading the passfile
+///  - `config` doesn't have a user set.
+///
+/// If `config` doesn't have a dbname set, matches the dbname field against the user name.
+///
+/// Intended to match libpq's behavior closely.
+///
+/// Unlike libpq, doesn't attempt to convert a `host` set to the default socket path into `"localhost"` for matching
+/// purposes. So passfiles using `localhost` to supply passwords for unix socket connections won't work.
+///
+/// Unlike libpq, doesn't enforce any rules on the passfile's permissions.
```

Comment:
Why not? This can be accessed through MetadataExt: https://doc.rust-lang.org/std/os/unix/fs/trait.MetadataExt.html#tymethod.mode

Reply:
It's easy to implement, but I'm not sure how to manage the integration tests: git can't directly represent a file with restricted permissions. I could have the tests create the files instead of shipping them in git, or have the tests change the permissions, or have the docker setup scripts install something suitable. What would you advise?

Reply:
I think I'd probably use the tempfile crate to make a temporary directory for the test and then make the passfile with appropriate permissions.

Reply:
Thanks, I've tried it that way and submitted it as #766